docs/Operators.md

```diff
@@ -7334,13 +7334,14 @@ expect(node, inputs=[x], outputs=[y], name="test_depthtospace_example")
 
 ### <a name="DequantizeLinear"></a><a name="dequantizelinear">**DequantizeLinear**</a>
 
-  The linear dequantization operator. It consumes a quantized tensor, a scale, and a zero point to compute the full precision tensor.
-  The dequantization formula is `y = (x - x_zero_point) * x_scale`. `x_scale` and `x_zero_point` must have same shape, and can be either a scalar
-  for per-tensor / per layer quantization, or a 1-D tensor for per-axis quantization.
-  `x_zero_point` and `x` must have same type. `x` and `y` must have same shape. In the case of dequantizing int32,
+  The linear dequantization operator. It consumes a quantized tensor, a scale, and a zero point to compute the full-precision tensor.
+  The dequantization formula is `y = (x - x_zero_point) * x_scale`. `x_scale` and `x_zero_point` must have the same shape, determining the quantization's granularity:
```

Comment:
How to determine the group size of the block quantization?

Reply:
> How to determine the group size of the block quantization?

And which axis is the blocked quantization applied to?

Reply:
The full description is in under `QuantizeLinear_ver21_doc`, and I've referred to it in here, because it seems redundant to repeat this for Dequantize. 

```
- Blocked quantization: The scale's shape is identical to the input's shape, except for one dimension, in which blocking is performed. For an input shape
  `(D0, ..., Di, ..., Dn)`, blocking along the i-th dimension with a block size `S`, `y_scale` shape is `(D0, ..., Di/S, ..., Dn)`. Blocked quantization
  is not supported for 1-D input tensors.
```
That is (pseudo code):
```block_shape = input.shape/scale.shape
block_axis = where(block_shape != 1)
block_size - block_shape[block_axis]
```

Reply:
```
For an input shape
  `(D0, ..., Di, ..., Dn)`, blocking along the i-th dimension with a block size `S`
  ```
  [source](https://github.com/galagam/onnx/blob/blocked-quantization/onnx/defs/quantization/defs.cc#L26)

Reply:
How S is calculated? Let's take Di equal to 15. And S = 32, 64 or 128 can make ceil(S/Di) to 1.

Reply:
S must divide Di without a remainder. I see that I left that out of the documentation. Thank you for noticing.

How about this, is that clear enough?
```
 - Blocked quantization: The scale's shape is identical to the input's shape, except for one dimension, in which blocking is performed. For an input shape
  `(D0, ..., Di, ..., Dn)`, blocking along the i-th dimension with a block size `S`, `y_scale` shape is `(D0, ..., Di/S, ..., Dn)`. The block size `S` must
  be a divisor of the input dimension Di. Blocked quantization is not supported for 1-D input tensors.
```

Reply:
It is a quite strong limit. How about to add an attribute to specify block(group) size?

Reply:
What is the use case you're targeting? Setting a block size that doesn't divide the input shape?
Both [AWQ (arxiv)](https://arxiv.org/abs/2306.00978) and [GPTQ (arxiv)](https://arxiv.org/abs/2210.17323) use small power-of-two block sizes, and the [OCP Microscaling Formats Spec v1.0](https://www.opencompute.org/documents/ocp-microscaling-formats-mx-v1-0-spec-final-pdf) went further and defined a single block size of 32.
In light of that, I think it's a reasonable constraint. It also also allows us to keep the API simple.
If there's a need, we can always add this attribute later on.

Reply:
It is fine to use small power-of-two for block size, but we cannot just assume the target tensor's shape is a multiplier of power-of-two. Just like what i mentioned above, we also need to support a matrix with shape [60, 128].

Reply:
The limit "The block size `S` must be a divisor of the input dimension Di." is unnecessary. AWQ and GPTQ are for LLM only, but the block-wise quantizelinear can be used on any model potentially.

Reply:
I think it's a question of generalizing vs. keeping a lean API.
Supporting a block size that doesn't divide the input in the blocking axis requires specifying the block size explicitly (i.e. by adding an attribute to QuantizeLinear and DequantizeLinear).
I'm fine with either decision, tagging ONNX maintainers to provide an input from their perspective @justinchuby @xadupre

Reply:
Adding an attribute is acceptable and seems reasonable if it supports other use-cases. But a quick question about the case where block-size S does not divide the dimension-size evenly: so, I guess the scale/zeropoint will have size rounded up (to ceiling (dim-size / S))?

Using a blocksize attribute (optional, used only for blocked quantization) also has another advantage: it clearly helps distinguish blocked quantization from others (since the attribute is statically available, while the input-dimension/rank/shape may sometimes not be known). This will be useful in version-adapter from 21 to 20 (since we should fail version-conversion for blocked-quantization).

Reply:
In that case, I suggest adding an attribute `block_size`.
Default value 0 means no blocking (per-tensor/per-axis quantization according to existing logic).
When `block_size` is provided, the blocking dimension is divided to blocks of size `block_size`, possibly with a smaller last block.

Reply:
@gramalingam @yufenglee please review my latest commit, adding the `block_size` attribute.